Fix ar command path in GHC

TerrorJack opened the pull request:
Previously, the "ar command" in the global config of GHC in nixpkgs is
simply "ar" instead of a proper absolute path in the nix store. This
will result in an "ar: command not found" error when using GHC and cabal
in a pure nix shell. This commit adds the patch and applies to all
pre-9.0 versions.

See output of ghc --info for "ar command" value.

The original patch comes from haskell.nix, see [here](https://github.com/input-output-hk/haskell.nix/blob/master/overlays/patches/ghc/respect-ar-path.patch). Hopefully this will be fixed in upstream 9.x releases.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.16.0-9-ga573266 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/de484d699eb79eab585ba992a763ae6776b65891.

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.16.0-9-ga573266 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/f9bd3516b69877f9879f189d86509633099f7dfc.

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.16.0-9-ga573266 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/c1d88e6debc32f5984294fa283dfb74255f8df88.

cdepillabout commented on pkgs/development/compilers/ghc/respect-ar-path.patch:
```diff
@@ -0,0 +1,25 @@
+diff -urd a/aclocal.m4 b/aclocal.m4
```
Generally we don't carry patches around in nixpkgs if we can avoid it.  Can you pull this from upstream with `fetchpatch`?

peti replied:
Yes, it would be good to know what the origin of that patch is and an easy way to document that is to just `fetchPatch` it from the origin.

TerrorJack replied:
We can't use the upstream patch in this case. The upstream patch only fixes `aclocal.m4`, but the source tarballs we fetch from upstream contains auto-generated `configure` script which also needs to be patched, so the in-tree patch needs to patch `configure` as well.

cdepillabout commented:
A couple questions:

- How can I test that this works?
- What does "not working" look like?
- Why does this need to be patched?

(Although I guess are all basically the same underlying question.)

xaverdh commented:
Nice, this should resolve https://github.com/NixOS/nixpkgs/issues/55995 if I am not mistaken?

TerrorJack committed:
Fix ar command path in GHC

Previously, the "ar command" in the global config of GHC in nixpkgs is
simply "ar" instead of a proper absolute path in the nix store. This
will result in an "ar: command not found" error when using GHC and cabal
in a pure nix shell. This commit adds the patch and applies to all
pre-9.0 versions.

See output of ghc --info for "ar command" value.

TerrorJack commented:
@cdepillabout 

> How can I test that this works? What does "not working" look like? Why does this need to be patched?

See #55995. tl;dr: `nix run -f /path/to/nixpkgs haskell.compiler.ghc8103 cabal-install`, then `cabal get unordered-containers-0.2.13.0 && cd unordered-containers-0.2.13.0 && cabal v2-build`. Before applying the fix, there'll be an error message like: `cabal: The program 'ar' is required but it could not be found.`.

> Can you rebase this on the haskell-updates branch as well?

Done. Should I change the PR to target `haskell-updates` instead of `master` as well?

@xaverdh

> Nice, this should resolve #55995 if I am not mistaken?

Yes.

cdepillabout commented:
I was able to reproduce this using the steps explained by @TerrorJack in https://github.com/NixOS/nixpkgs/pull/110695#issuecomment-766693562.  One thing that I had to be careful with is that I have `ar` available in `/run/current-system/sw/bin/ar`, so I had to make sure to run `nix run --ignore-environment -f ../. cabal-install ghc -c cabal build` (specifying the `--ignore-environment` flag).

On https://github.com/NixOS/nixpkgs/pull/85136#issuecomment-620530729, @peti specifically talks about the case where you can get `cabal-install` from nixpkgs, but get binutils and ghc from your distro.  However, I don't really understand whether the fix in this PR is fixing a bug of GHC (in that the path to `ar` is not properly recorded) or a "feature" of GHC (in that you can use a different version of binutils when actually using GHC).  If it is fixing a bug, then I imagine it should probably be fixed upstream?  Is there a bug report upstream somewhere?  If not, could you create one @TerrorJack?

TerrorJack commented:
@cdepillabout https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4885

cdepillabout commented:
@TerrorJack Can you edit this PR to add a comment to link to the upstream PR, and add an explanation of why we can't just pull this from upstream?

I'd like to make sure we can track when this patch is no longer needed (so we know when it can be removed from nixpkgs).

TerrorJack committed:
Add comments

cdepillabout commented:
@GrahamcOfBorg build haskell.compilers.ghc865
@GrahamcOfBorg build haskell.compilers.ghc882
@GrahamcOfBorg build haskell.compilers.ghc883
@GrahamcOfBorg build haskell.compilers.ghc884
@GrahamcOfBorg build haskell.compilers.ghc8101
@GrahamcOfBorg build haskell.compilers.ghc8102
@GrahamcOfBorg build haskell.compilers.ghc8103

cdepillabout reviewed (approved):
This looks good to me, especially since it seems like they are accepting it as a bug upstream.

I'll wait a week for anyone (like @peti) to comment, but if no one has said anything by then, please feel free to ping me again @TerrorJack and I'll merge this in.

peti reviewed (approved):
I think this looks pretty good. Thanks a lot to both of you for your efforts! :+1:

Let's wait for the build bot to confirm that everything compiles, and then we should merge so that these patches make into this weeks merge of `haskell-updates` into `master`.

peti commented:
Well, the test builds don't look so good. That patch is probably not going to make it into `haskell-updates` tonight. :-(

peti commented:
I've pushed 8fae904eb6ac3e5cb863f9679d2190652686e3b5 to `haskell-updates`. Let's hope that the test builds succeed on Hydra at https://hydra.nixos.org/jobset/nixpkgs/haskell-updates.

cdepillabout commented:
This was merged into `haskell-updates` in 643169bbb48126a3b6ec3c4a32bbbe7a815d11c8 (and is now in `master` as well).

Thanks @TerrorJack!
